### PR TITLE
Tooling, Security and Build Overhaul

### LINKED ISSUE #304: Improving Security Tooling and Transparency
**Describe what problem your feature request solves**
ThreatDragon should be clear and transparent on what tooling/automation is in place that provides any level of security.

**Describe the solution you'd like**
Update the `Contributing.md` file to include build processes, required checks/tooling.

**Additional context**
This was brought up in a discussion in our Slack channel (feel free to drop by #project-threat-dragon in the OWASP Slack!)
I'd love some feedback/discussion/suggestions on this!

Here's what we have currently, or plan to have soon:
|Type|Tool|Version 1.x|Version 2.x|
|---|---|---|--|
|SAST|[LGTM](https://lgtm.com/)|x|x|
|DAST|[OWASP ZAP](https://www.zaproxy.org/)||x|
|Container Scanning|[trivy](https://github.com/aquasecurity/trivy)||x|
|SCA|[trivy](https://github.com/aquasecurity/trivy)||x|

This will be hosted in the documentation: https://www.threatdragon.com/docs/ under the "trust" dropdown

### PR DESCRIPTION
**Pull request by lreading**
**Summary**
Large overhaul of tooling, security and build process.

**Description for the changelog**
Closes #304 
Updated to use node 14
Now using pnpm
Refactor dockerfile
Documentation
Trivy ran per commit

**Node 16**
The audit capabilities on version 16 of node have been a bit more reliable, at least from my own experience.  Since we're making such big/sweeping changes, I think it's a good time to make the upgrade.  td.server and td.site didn't seem to mind running on 16.

**pnpm**
Pnpm makes package management a lot easier.  If you're in a dependency nightmare (especially with a vulnerable library), adding one line to the package.json will force a resolution to a specific version.  For example:
```
    ...
    "pnpm": {
        "overrides": {
            "ansi-regex@>2.1.1 <5.0.1": ">=5.0.1"
             ...
        }
    },
    ...
```
There is a slight learning curve to it, but is relatively easy to pick up if you're already familiar with npm.  I've also added a pre-install command to prevent anyone from running `npm install`.  Some other highlights:
- `pnpm install` Same as npm install (-r for recursive, prod and dev flags available as well)
- `pnpm add <package_name>` Adds a package as a *dependency*
- `pnpm add -D <package_name>` Adds a *dev* dependency
- `pnpm audit` Same as npm audit
- `pnpm run <script_name>` Runs a script as defined in the package.json file

You can still run scripts using `npm run <script>` without issue.  pnpm is only _required_ for package management.

**Trivy Integration**
Trivy will now be run on each commit.  It scans the image and packages contained within the application.  There is a `.trivyignore` file for ignoring CVEs if appropriate, but we should strive to avoid that without strong justification. Of course, as I write this, there's already one in there for a CVE that's baked into the nodejs alpine image.

There's also a cron-based GH Action that should run daily against v2-development to check for new vulnerabilities.  If any vulnerabilities are discovered via the cron, an alert is created in the security tab for this repo.  Because it's not currently the default branch, I don't believe we'll get emails or notifications for it.  After v2 becomes the default, it should all work as expected.  This should keep us on top of emerging vulnerabilities should the repo go some time without activity.

**Dockerfile rewrite**
We are no longer inheriting from the distroless.  There were multiple packages with vulnerabilities in the version we were using, and no way of managing the packages in that image.  We are now using node:16-alpine and updating dependencies each time we build.  We are running as the node user.

While the build didn't get any faster from this work, it definitely helps us keep Threat Dragon more secure.  Here's a screenshot of the Trivy results from the latest v2-development image:
Screenshot from 2021-12-19 03-18-16
Screenshot from 2021-12-19 03-18-23

________________________ 

Here is the scan of the image built with the changes from this PR (ignoring CVE-2021-3807 for now):
Screenshot from 2021-12-19 03-20-50

### CODE REVIEW

**Comment by lreading** [comment]
Please take your time with this, and only when you have a chance to look.  It's a lot of changes all at once. :smiley:

**Comment by jgadsden** [comment]
I get an error when running the application : "vue-cli-service: command not found". Attached is my console log, clone and install seems successful, just `npm start`

@lreading I can have a look at this if you like?

[console.txt](https://github.com/OWASP/threat-dragon/files/7784485/console.txt)

**Comment by lreading** [comment]
> I get an error when running the application : "vue-cli-service: command not found". Attached is my console log, clone and install seems successful, just `npm start`
> 
> @lreading I can have a look at this if you like? Probably just needs to be added to `td.vue/package.json`
> 
> [console.txt](https://github.com/OWASP/threat-dragon/files/7784485/console.txt)

I'll do a fresh pull somewhere and check it out. :) 

Did you do a `pnpm install` from the root, or in `td.vue`?

**Review by jgadsden — APPROVED** [review]
Looks good to me as something like `postinstall` is being added to `package.json`
Thanks @lreading for doing all this

File: docs/development/local.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+1. Run `pnpm install`
+1. Configure your [environment](./env)
+1. Run `npm run dev:server`
+1. In another terminal, run `npm run dev:vue`
+1. Visit [http://localhost:8080](http://localhost:8080/)
+
+## Dependencies
+
+### Pnpm
+Threat Dragon uses pnpm as a package manager.  You can install it by using `npm`, or by following the [installation instructions](https://pnpm.io/installation) on their website.
```
[the remark this answers is not in the document]
Comment (lreading) [reply docs/development/local.md]: pnpm can actually install node for you as well, so it becomes a chicken/egg thing. :) That said, I 100% agree.  I highly doubt anyone will be working on threat dragon who doesn't already have node installed. I've updated it to first suggest `npm install -g pnpm`, but still includes the link for "alternative installation methods",  if you think that would work?

**Comment by lreading** [comment]
I added the postinstall script to ensure all subproject dependencies are installed when running `pnpm install` from the root of the repo.  I also updated the `start` command to run both td.server and td.vue in watch mode, as both projects now havve the capability.

**Comment by jgadsden** [comment]
I ran the updated file set and all works for me now - thanks @lreading for sorting this out 👍🏾

**Pull request merged**